// @clawdbot/lobster/dist/src/pipeline_resume_state.js
import { randomUUID } from 'node:crypto';
import { encodeToken } from './token.js';
import { cleanupApprovalIndexByStateKey, createApprovalIndex, deleteStateJson, readStateJson, writeStateJson, } from './state/store.js';
import { sharedAjv } from './validation.js';
export function extractPipelineHalt(output) {
    const halted = output.halted && output.items.length === 1
        ? output.items[0]
        : null;
    const approval = halted?.type === 'approval_request'
        ? halted
        : null;
    const inputRequest = halted?.type === 'input_request'
        ? halted
        : null;
    return { approval, inputRequest };
}
export async function finalizePipelineToolRun(params) {
    const { approval, inputRequest } = extractPipelineHalt(params.output);
    if (approval) {
        const nextStateKey = await savePipelineResumeState(params.env, {
            pipeline: params.pipeline,
            resumeAtIndex: (params.output.haltedAt?.index ?? -1) + 1,
            items: approval.items,
            haltType: 'approval_request',
            prompt: approval.prompt,
            createdAt: new Date().toISOString(),
        });
        if (params.previousStateKey) {
            await cleanupApprovalIndexByStateKey({ env: params.env, stateKey: params.previousStateKey });
            await deleteStateJson({ env: params.env, key: params.previousStateKey });
        }
        let approvalId;
        try {
            approvalId = await createApprovalIndex({ env: params.env, stateKey: nextStateKey });
        }
        catch (err) {
            await deleteStateJson({ env: params.env, key: nextStateKey }).catch(() => { });
            throw err;
        }
        const resumeToken = encodeToken({
            protocolVersion: 1,
            v: 1,
            kind: 'pipeline-resume',
            stateKey: nextStateKey,
        });
        return {
            status: 'needs_approval',
            output: [],
            requiresApproval: {
                ...approval,
                resumeToken,
                approvalId,
            },
            requiresInput: null,
        };
    }
    if (inputRequest) {
        const nextStateKey = await savePipelineResumeState(params.env, {
            pipeline: params.pipeline,
            resumeAtIndex: (params.output.haltedAt?.index ?? -1) + 1,
            items: [],
            haltType: 'input_request',
            inputSchema: inputRequest.responseSchema,
            prompt: inputRequest.prompt,
            createdAt: new Date().toISOString(),
        });
        if (params.previousStateKey) {
            await cleanupApprovalIndexByStateKey({ env: params.env, stateKey: params.previousStateKey });
            await deleteStateJson({ env: params.env, key: params.previousStateKey });
        }
        const resumeToken = encodeToken({
            protocolVersion: 1,
            v: 1,
            kind: 'pipeline-resume',
            stateKey: nextStateKey,
        });
        return {
            status: 'needs_input',
            output: [],
            requiresApproval: null,
            requiresInput: {
                type: 'input_request',
                prompt: inputRequest.prompt,
                responseSchema: inputRequest.responseSchema,
                ...(inputRequest.defaults !== undefined ? { defaults: inputRequest.defaults } : null),
                ...(inputRequest.subject !== undefined ? { subject: inputRequest.subject } : null),
                resumeToken,
            },
        };
    }
    if (params.previousStateKey) {
        await cleanupApprovalIndexByStateKey({ env: params.env, stateKey: params.previousStateKey });
        await deleteStateJson({ env: params.env, key: params.previousStateKey });
    }
    return {
        status: 'ok',
        output: params.output.items,
        requiresApproval: null,
        requiresInput: null,
    };
}
export async function savePipelineResumeState(env, state) {
    const stateKey = `pipeline_resume_${randomUUID()}`;
    await writeStateJson({ env, key: stateKey, value: state });
    return stateKey;
}
export async function loadPipelineResumeState(env, stateKey) {
    const stored = await readStateJson({ env, key: stateKey });
    if (!stored || typeof stored !== 'object') {
        throw new Error('Pipeline resume state not found');
    }
    const data = stored;
    if (!Array.isArray(data.pipeline))
        throw new Error('Invalid pipeline resume state');
    if (typeof data.resumeAtIndex !== 'number')
        throw new Error('Invalid pipeline resume state');
    if (!Array.isArray(data.items))
        throw new Error('Invalid pipeline resume state');
    if (data.haltType !== undefined && !['approval_request', 'input_request'].includes(data.haltType)) {
        throw new Error('Invalid pipeline resume state');
    }
    return data;
}
export function validatePipelineInputResponse(schema, response) {
    if (schema === undefined) {
        throw new Error('pipeline input response schema is missing');
    }
    let validator;
    try {
        validator = sharedAjv.compile(schema);
    }
    catch {
        throw new Error('pipeline input response schema is invalid');
    }
    const ok = validator(response);
    if (ok)
        return;
    const first = validator.errors?.[0];
    const pathValue = first?.instancePath || '/';
    const reason = first?.message ? ` ${first.message}` : '';
    throw new Error(`pipeline input response failed schema validation at ${pathValue}:${reason}`);
}
