// @buape/carbon/dist/src/internals/BoundedExecutor.js
export function createBoundedExecutor({ concurrency, run, getQueuedAt }) {
    const queue = [];
    let inFlight = 0;
    const idleWaiters = [];
    const notifyIfIdle = () => {
        if (inFlight > 0 || queue.length > 0)
            return;
        for (const resolve of idleWaiters.splice(0)) {
            resolve();
        }
    };
    const pump = () => {
        while (inFlight < concurrency && queue.length > 0) {
            const task = queue.shift();
            if (!task)
                continue;
            inFlight += 1;
            void run(task)
                .catch(() => {
                // caller handles errors inside run
            })
                .finally(() => {
                inFlight -= 1;
                pump();
                notifyIfIdle();
            });
        }
    };
    return {
        schedule(task) {
            queue.push(task);
            pump();
        },
        async onIdle() {
            if (inFlight === 0 && queue.length === 0)
                return;
            await new Promise((resolve) => {
                idleWaiters.push(resolve);
            });
        },
        getInFlight() {
            return inFlight;
        },
        getQueueDepth() {
            return queue.length;
        },
        getOldestQueuedAgeMs() {
            if (queue.length === 0 || !getQueuedAt)
                return 0;
            const first = queue[0];
            if (!first)
                return 0;
            const queuedAt = getQueuedAt(first);
            if (typeof queuedAt !== "number")
                return 0;
            return Math.max(0, Date.now() - queuedAt);
        }
    };
}
