// @microsoft/teams.apps/dist/test-utils.js
"use strict";
/**
 * Test utilities for creating Apps in test environments
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestAdapter = void 0;
exports.createTestApp = createTestApp;
const app_1 = require("./app");
/**
 * Mock HTTP adapter for testing
 * Provides no-op implementations for all methods
 */
class TestAdapter {
    registerRoute(_method, _path, _handler) {
        // No-op for tests
    }
    serveStatic(_path, _directory) {
        // No-op for tests
    }
    async start(_port) {
        // No-op for tests
    }
    async stop() {
        // No-op for tests
    }
}
exports.TestAdapter = TestAdapter;
/**
 * Creates an App instance configured for testing
 * Automatically uses TestAdapter if no httpServerAdapter is provided
 *
 * @param options App configuration options
 * @returns App instance with TestAdapter
 *
 * @example
 * const app = createTestApp({
 *   clientId: 'test-client-id',
 *   clientSecret: 'test-client-secret'
 * });
 */
function createTestApp(options) {
    return new app_1.App({
        ...options,
        httpServerAdapter: options?.httpServerAdapter ?? new TestAdapter()
    });
}
