// @microsoft/teams.apps/dist/http/express-adapter.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExpressAdapter = void 0;
const http_1 = __importDefault(require("http"));
const cors_1 = __importDefault(require("cors"));
const express_1 = __importDefault(require("express"));
const teams_common_1 = require("@microsoft/teams.common");
/**
 * Express adapter for HttpServer
 *
 * Handles Express-specific HTTP framework concerns:
 * - Express app creation and middleware setup
 * - Route registration via Express routing
 * - Request/response data extraction and sending
 * - Server lifecycle management
 */
class ExpressAdapter {
    // Expose Express methods for backwards compatibility
    get;
    post;
    patch;
    put;
    delete;
    route;
    use;
    express;
    server;
    logger;
    onError;
    constructor(serverOrApp, options) {
        if (serverOrApp instanceof http_1.default.Server) {
            // The adapter handles all requests on this server. Use the adapter's
            // methods (get, post, use, etc.) to add routes. If you need your own
            // Express app, pass it in instead and manage the server yourself.
            this.express = (0, express_1.default)();
            this.server = serverOrApp;
            this.server.on('request', this.express);
        }
        else if (typeof serverOrApp === 'function') {
            this.express = serverOrApp;
        }
        else {
            this.express = (0, express_1.default)();
            this.server = http_1.default.createServer(this.express);
        }
        this.logger = options?.logger ?? new teams_common_1.ConsoleLogger('ExpressAdapter');
        this.onError = options?.onError;
        // Bind Express methods
        this.get = this.express.get.bind(this.express);
        this.post = this.express.post.bind(this.express);
        this.patch = this.express.patch.bind(this.express);
        this.put = this.express.put.bind(this.express);
        this.delete = this.express.delete.bind(this.express);
        this.route = this.express.route.bind(this.express);
        this.use = this.express.use.bind(this.express);
    }
    /**
     * Register a route handler for a given HTTP method and path
     */
    registerRoute(method, path, handler) {
        if (method !== 'POST') {
            throw new Error(`Unsupported HTTP method: ${method}`);
        }
        const expressHandler = async (req, res, next) => {
            try {
                const response = await handler({
                    body: req.body,
                    headers: req.headers
                });
                res.status(response.status).send(response.body);
            }
            catch (err) {
                next(err);
            }
        };
        this.express.post(path, express_1.default.json(), expressHandler);
    }
    /**
     * Start the server listening on the specified port
     */
    async start(port) {
        const server = this.server;
        if (!server) {
            throw new Error('Cannot start: server lifecycle is managed externally. Call listen() on your Express app or http.Server directly.');
        }
        return new Promise((resolve, reject) => {
            // Handle startup errors
            server.once('error', (err) => {
                if (this.onError) {
                    this.onError(err);
                }
                reject(err);
            });
            server.listen(port, () => {
                this.logger.info(`listening on port ${port} 🚀`);
                // Set up persistent error listener after startup
                if (this.onError) {
                    server.on('error', this.onError);
                }
                resolve();
            });
        });
    }
    /**
     * Serve static files from a directory
     */
    serveStatic(path, directory) {
        this.express.use(path, (0, cors_1.default)(), express_1.default.static(directory));
    }
    /**
     * Stop the server and close all connections
     */
    async stop() {
        const server = this.server;
        if (!server) {
            throw new Error('Cannot stop: server lifecycle is managed externally. Call close() on your Express app or http.Server directly.');
        }
        return new Promise((resolve, reject) => {
            server.close((err) => {
                if (err) {
                    reject(err);
                }
                else {
                    this.logger.info('server stopped');
                    resolve();
                }
            });
        });
    }
}
exports.ExpressAdapter = ExpressAdapter;
