// @buape/carbon/dist/src/plugins/cloudflare-gateway/CloudflareGatewayPlugin.js
import { Plugin } from "../../abstracts/Plugin.js";
const defaults = {
    forwardPath: "/__carbon/gateway-events",
    secretHeader: "x-carbon-gateway-secret",
    durableObjectBinding: "CARBON_GATEWAY",
    durableObjectName: "default",
    connectPath: "/connect",
    connectDebounceMs: 10_000
};
export class CloudflareGatewayPlugin extends Plugin {
    id = "cloudflare-gateway";
    options;
    client;
    lastConnectAttempt = 0;
    constructor(options) {
        super();
        this.options = {
            ...defaults,
            ...options
        };
    }
    registerClient(client) {
        this.client = client;
        // Workers + forwarded gateway listeners should avoid cross-request scheduler promises.
        client.rest.options.queueRequests = false;
    }
    registerRoutes(client) {
        client.routes.push({
            method: "POST",
            path: this.options.forwardPath,
            handler: this.handleForwardRequest.bind(this)
        });
    }
    onRequest(req, ctx) {
        if (!this.client) {
            return;
        }
        const now = Date.now();
        if (now - this.lastConnectAttempt < this.options.connectDebounceMs) {
            return;
        }
        const url = new URL(req.url);
        if (url.pathname === this.options.forwardPath) {
            return;
        }
        const envValue = ctx && typeof ctx === "object" ? Reflect.get(ctx, "env") : null;
        if (!envValue || typeof envValue !== "object") {
            return;
        }
        const env = envValue;
        this.lastConnectAttempt = now;
        const connectPromise = this.connectDurableObject(env).catch((error) => {
            console.error("[CloudflareGatewayPlugin] Failed to connect durable object", error);
        });
        ctx.waitUntil?.(connectPromise);
    }
    async handleForwardRequest(req) {
        if (!this.client) {
            return new Response("Service Unavailable", { status: 503 });
        }
        const deploySecret = typeof process === "undefined" ? "" : (process.env?.DEPLOY_SECRET ?? "");
        if (req.headers.get(this.options.secretHeader) !== deploySecret) {
            return new Response("Unauthorized", { status: 401 });
        }
        const payload = (await req.json());
        if (!payload.type || !payload.data) {
            return new Response("Bad Request", { status: 400 });
        }
        const type = payload.type;
        const rawPayload = {
            ...payload.data,
            clientId: this.client.options.clientId
        };
        const accepted = this.client.eventHandler.handleEvent(rawPayload, type);
        if (!accepted) {
            return new Response("Service Unavailable", { status: 503 });
        }
        return new Response(null, { status: 204 });
    }
    async connectDurableObject(env) {
        const namespaceValue = env[this.options.durableObjectBinding];
        if (!namespaceValue || typeof namespaceValue !== "object") {
            return;
        }
        const namespaceMethods = namespaceValue;
        if (typeof namespaceMethods.idFromName !== "function" ||
            typeof namespaceMethods.get !== "function") {
            return;
        }
        const processEnv = typeof process === "undefined" ? undefined : process.env;
        const baseUrl = processEnv?.BASE_URL ?? env.BASE_URL ?? "";
        const deploySecret = processEnv?.DEPLOY_SECRET ??
            env.DEPLOY_SECRET ??
            "";
        const token = processEnv?.DISCORD_BOT_TOKEN ??
            env.DISCORD_BOT_TOKEN ??
            "";
        const clientId = processEnv?.DISCORD_CLIENT_ID ??
            env.DISCORD_CLIENT_ID ??
            "";
        if (!baseUrl || !deploySecret || !token || !clientId) {
            return;
        }
        const config = {
            baseUrl,
            deploySecret,
            token,
            clientId,
            forwardPath: this.options.forwardPath,
            secretHeader: this.options.secretHeader,
            gateway: {
                intents: this.options.intents,
                autoInteractions: this.options.autoInteractions,
                reconnect: this.options.reconnect
            }
        };
        const namespace = namespaceValue;
        const id = namespace.idFromName(this.options.durableObjectName);
        const stub = namespace.get(id);
        await stub.fetch(`https://carbon.gateway${this.options.connectPath}`, {
            method: "POST",
            headers: {
                "content-type": "application/json"
            },
            body: JSON.stringify(config)
        });
    }
}
