// @microsoft/teams.apps/dist/types/plugin/decorators/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Event = exports.HttpServer = exports.Storage = exports.Logger = exports.Dependency = exports.Plugin = void 0;
var plugin_1 = require("./plugin");
Object.defineProperty(exports, "Plugin", { enumerable: true, get: function () { return plugin_1.Plugin; } });
var dependency_1 = require("./dependency");
Object.defineProperty(exports, "Dependency", { enumerable: true, get: function () { return dependency_1.Dependency; } });
Object.defineProperty(exports, "Logger", { enumerable: true, get: function () { return dependency_1.Logger; } });
Object.defineProperty(exports, "Storage", { enumerable: true, get: function () { return dependency_1.Storage; } });
Object.defineProperty(exports, "HttpServer", { enumerable: true, get: function () { return dependency_1.HttpServer; } });
var event_1 = require("./event");
Object.defineProperty(exports, "Event", { enumerable: true, get: function () { return event_1.Event; } });
