// @microsoft/teams.apps/dist/middleware/auth/remote-function-validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RemoteFunctionValidator = void 0;
const jwt_validator_1 = require("./jwt-validator");
/**
 * Remote function validator for /api/functions/* requests
 * Wraps an Entra token validator and provides a simple check() method
 */
class RemoteFunctionValidator {
    entraTokenValidator;
    logger;
    constructor(entraTokenValidator, logger) {
        this.entraTokenValidator = entraTokenValidator;
        this.logger = logger;
    }
    /**
     * Create a remote function validator for Entra tokens
     */
    static create(tenantId, clientId, logger, options) {
        const jwtValidator = new jwt_validator_1.JwtValidator({
            clientId,
            tenantId,
            validateIssuer: {
                allowedTenantIds: options?.allowedTenantIds
            },
            validateScope: options?.requiredScope ? { requiredScope: options.requiredScope } : undefined,
            jwksUriOptions: {
                type: 'tenantId'
            }
        }, logger);
        return new RemoteFunctionValidator(jwtValidator, logger);
    }
    async check(headers) {
        const h = (key) => {
            const v = headers[key];
            return Array.isArray(v) ? v[0] : v;
        };
        const appSessionId = h('x-teams-app-session-id');
        const pageId = h('x-teams-page-id');
        const authorization = h('authorization')?.split(' ');
        const authToken = authorization?.length === 2 && authorization[0].toLowerCase() === 'bearer'
            ? authorization[1]
            : '';
        const tokenPayload = await this.entraTokenValidator.validateAccessToken(authToken);
        if (!pageId ||
            !appSessionId ||
            !authToken ||
            !tokenPayload) {
            this.logger.debug('unauthorized - missing required headers or invalid token');
            return null;
        }
        return {
            appId: tokenPayload?.['appId'],
            appSessionId,
            authToken,
            channelId: h('x-teams-channel-id'),
            chatId: h('x-teams-chat-id'),
            meetingId: h('x-teams-meeting-id'),
            messageId: h('x-teams-message-id'),
            pageId,
            subPageId: h('x-teams-sub-page-id'),
            teamId: h('x-teams-team-id'),
            tenantId: tokenPayload['tid'],
            userId: tokenPayload['oid'],
            userName: tokenPayload['name'],
        };
    }
}
exports.RemoteFunctionValidator = RemoteFunctionValidator;
