// @microsoft/teams.apps/dist/middleware/auth/service-token-validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServiceTokenValidator = void 0;
const jwt_validator_1 = require("./jwt-validator");
/**
 * Service token validator for Bot Framework /api/messages requests
 * Validates Bot Framework service tokens
 */
class ServiceTokenValidator {
    jwtValidator;
    credentials;
    constructor(appId, tenantId, serviceUrl, logger) {
        this.jwtValidator = new jwt_validator_1.JwtValidator({
            clientId: appId,
            tenantId,
            validateIssuer: { allowedIssuer: 'https://api.botframework.com' },
            validateServiceUrl: serviceUrl ? { expectedServiceUrl: serviceUrl } : undefined,
            jwksUriOptions: {
                type: 'uri',
                uri: 'https://login.botframework.com/v1/.well-known/keys'
            },
        }, logger);
        this.credentials = { clientId: appId, tenantId };
    }
    async check(authHeader, body) {
        // Extract token from "Bearer <token>" format
        const token = authHeader.startsWith('Bearer ')
            ? authHeader.substring(7)
            : authHeader;
        // Validate token using JWT validator
        const payload = await this.jwtValidator.validateAccessToken(token, {
            validateServiceUrl: body.serviceUrl ? { expectedServiceUrl: body.serviceUrl } : undefined
        });
        if (!payload) {
            throw new Error('Invalid token');
        }
        // Convert JWT payload to IToken
        return {
            appId: payload.appid || this.credentials?.clientId || '',
            from: 'azure',
            fromId: payload.sub || '',
            serviceUrl: body.serviceUrl || payload.serviceurl || '',
            isExpired: () => false, // Already validated by JWT validator
        };
    }
}
exports.ServiceTokenValidator = ServiceTokenValidator;
