// @microsoft/teams.cards/dist/actions/submit/collab-stage.mjs
import { SubmitAction, SubmitActionData, InvokeSubmitActionData, CollabStageInvokeDataValue } from '../../core';

class CollabStageAction extends SubmitAction {
  /**
   * Initial data that input fields will be combined with. These are essentially 'hidden' properties.
   */
  data;
  constructor(tab, options = {}) {
    super(options);
    Object.assign(this, options);
    this.data = new SubmitActionData({
      msteams: new InvokeSubmitActionData(
        tab ? new CollabStageInvokeDataValue({ tabInfo: tab }) : void 0
      )
    });
  }
  static from(options) {
    const msteams = options.data.msteams;
    const value = msteams?.value;
    return new CollabStageAction(value?.tabInfo, options);
  }
  withData(value) {
    super.withData(new SubmitActionData({ msteams: value }));
    return this;
  }
  withValue(value) {
    const msteams = this.data.msteams;
    if (msteams) {
      msteams.value = new CollabStageInvokeDataValue({ tabInfo: value });
    }
    return this;
  }
}

export { CollabStageAction };
