// @microsoft/teams.apps/dist/plugins/http/plugin.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpPlugin = void 0;
const http_1 = __importDefault(require("http"));
const cors_1 = __importDefault(require("cors"));
const express_1 = __importDefault(require("express"));
const package_json_1 = __importDefault(require("../../../package.json"));
const http_2 = require("../../http");
const http_server_1 = require("../../http/http-server");
const types_1 = require("../../types");
/**
 * @deprecated Use the httpServerAdapter option instead:
 *   new App({ httpServerAdapter: new ExpressAdapter() })
 *
 * This wrapper will be removed in a future version.
 *
 * NOTE: This plugin is named "HttpPlugin" for historical reasons and backwards compatibility.
 * It is the default HTTP plugin that uses Express as the underlying framework.
 */
let HttpPlugin = class HttpPlugin {
    // Expose Express methods for backwards compatibility
    get;
    post;
    patch;
    put;
    delete;
    route;
    use;
    expressAdapter;
    server;
    constructor(server, options) {
        this.expressAdapter = new http_2.ExpressAdapter(server);
        this.server = new http_server_1.HttpServer(this.expressAdapter, {
            ...options,
            messagingEndpoint: options?.messagingEndpoint ?? '/api/messages',
        });
        // Expose Express methods
        this.get = this.expressAdapter.get;
        this.post = this.expressAdapter.post;
        this.patch = this.expressAdapter.patch;
        this.put = this.expressAdapter.put;
        this.delete = this.expressAdapter.delete;
        this.route = this.expressAdapter.route;
        this.use = this.expressAdapter.use;
    }
    /**
     * App.constructor calls this to extract the HttpServer
     * @internal
     */
    asServer() {
        return this.server;
    }
    /**
     * Plugin lifecycle hook
     */
    async onInit() {
        // TODO: Setting cors globally and body parsing for all routes in /api
        // is actually a mistake. When HttpPlugin is officially deprecated, this
        // behavior will go away as well.
        this.use((0, cors_1.default)());
        this.use('/api', express_1.default.json());
    }
    /**
     * serve static files
     * @param path the url path to serve
     * @param dist the dist file path to serve
     */
    static(path, dist) {
        this.expressAdapter.serveStatic(path, dist);
        return this;
    }
};
exports.HttpPlugin = HttpPlugin;
exports.HttpPlugin = HttpPlugin = __decorate([
    (0, types_1.Plugin)({
        name: 'http',
        version: package_json_1.default.version,
        description: 'Deprecated: Use httpServerAdapter option instead',
    }),
    __metadata("design:paramtypes", [http_1.default.Server, Object])
], HttpPlugin);
