// skillflag/dist/core/list.js
import fs from "node:fs/promises";
import path from "node:path";
import { collectSkillEntries, createTarStream } from "./tar.js";
import { digestStreamSha256 } from "./digest.js";
import { listSkillDirs } from "./paths.js";
import { parseFrontmatter } from "../shared/frontmatter.js";
async function readSkillInfo(dir, id) {
    const skillMdPath = path.join(dir, "SKILL.md");
    try {
        const content = await fs.readFile(skillMdPath, "utf8");
        const fields = parseFrontmatter(content);
        const summary = fields.description
            ? fields.description.replace(/[\t\n]/g, " ").trim()
            : undefined;
        return {
            id,
            dir,
            summary,
            version: fields.version,
        };
    }
    catch {
        return { id, dir };
    }
}
export async function listSkills(rootDirs) {
    const seen = new Map();
    for (const rootDir of rootDirs) {
        const dirs = await listSkillDirs(rootDir);
        for (const skill of dirs) {
            if (!seen.has(skill.id)) {
                seen.set(skill.id, skill.dir);
            }
        }
    }
    const infos = [];
    for (const [id, dir] of seen.entries()) {
        infos.push(await readSkillInfo(dir, id));
    }
    infos.sort((a, b) => a.id.localeCompare(b.id));
    return infos;
}
export async function listSkillsJson(rootDirs) {
    const skills = await listSkills(rootDirs);
    const results = [];
    for (const skill of skills) {
        const { entries, fileCount } = await collectSkillEntries(skill.dir, skill.id);
        const stream = createTarStream(entries);
        const digest = await digestStreamSha256(stream);
        const item = {
            id: skill.id,
            digest,
        };
        if (fileCount > 0) {
            item.files = fileCount;
        }
        if (skill.summary) {
            item.summary = skill.summary;
        }
        if (skill.version) {
            item.version = skill.version;
        }
        results.push(item);
    }
    return {
        skillflag_version: "0.1",
        skills: results,
    };
}
