// skillflag/dist/core/tar.js
import fs from "node:fs";
import fsPromises from "node:fs/promises";
import path from "node:path";
import * as tar from "tar-stream";
import { SkillflagError } from "./errors.js";
const FIXED_MTIME = new Date(0);
function isInvalidRelPath(relPosix) {
    if (path.posix.isAbsolute(relPosix))
        return true;
    const parts = relPosix.split("/");
    return parts.includes("..");
}
async function collectEntriesForDir(rootDir, relPosix, id, dirs, files) {
    dirs.add(relPosix);
    const absDir = relPosix
        ? path.join(rootDir, ...relPosix.split("/"))
        : rootDir;
    const dirents = await fsPromises.readdir(absDir, { withFileTypes: true });
    for (const dirent of dirents) {
        const name = dirent.name;
        const relChild = relPosix ? `${relPosix}/${name}` : name;
        if (isInvalidRelPath(relChild)) {
            throw new SkillflagError(`Invalid path in skill: ${id}/${relChild}`);
        }
        const absChild = path.join(absDir, name);
        if (dirent.isDirectory()) {
            await collectEntriesForDir(rootDir, relChild, id, dirs, files);
            continue;
        }
        if (dirent.isFile()) {
            const stat = await fsPromises.stat(absChild);
            files.push({
                name: `${id}/${relChild}`,
                type: "file",
                absPath: absChild,
                size: stat.size,
                mode: stat.mode & 0o777,
            });
            continue;
        }
        if (dirent.isSymbolicLink()) {
            throw new SkillflagError(`Symlinks are not supported in skill bundles: ${id}/${relChild}`);
        }
        throw new SkillflagError(`Unsupported file type in skill bundle: ${id}/${relChild}`);
    }
}
export async function collectSkillEntries(skillDir, id) {
    const dirs = new Set();
    const files = [];
    await collectEntriesForDir(skillDir, "", id, dirs, files);
    const dirEntries = [];
    for (const relDir of dirs) {
        const absDir = relDir
            ? path.join(skillDir, ...relDir.split("/"))
            : skillDir;
        const stat = await fsPromises.stat(absDir);
        const dirName = relDir ? `${id}/${relDir}/` : `${id}/`;
        dirEntries.push({
            name: dirName,
            type: "directory",
            mode: stat.mode & 0o777,
        });
    }
    const entries = [...dirEntries, ...files].sort((a, b) => a.name.localeCompare(b.name));
    return { entries, fileCount: files.length };
}
function writeDirEntry(pack, entry) {
    return new Promise((resolve, reject) => {
        pack.entry({
            name: entry.name,
            type: "directory",
            mode: entry.mode,
            mtime: FIXED_MTIME,
            uid: 0,
            gid: 0,
            uname: "",
            gname: "",
        }, (err) => {
            if (err) {
                reject(err);
                return;
            }
            resolve();
        });
    });
}
function writeFileEntry(pack, entry) {
    if (!entry.absPath) {
        return Promise.reject(new Error(`Missing file path for ${entry.name}`));
    }
    if (entry.size === undefined) {
        return Promise.reject(new Error(`Missing file size for ${entry.name}`));
    }
    const absPath = entry.absPath;
    return new Promise((resolve, reject) => {
        const fileStream = fs.createReadStream(absPath);
        const tarEntry = pack.entry({
            name: entry.name,
            type: "file",
            mode: entry.mode,
            size: entry.size,
            mtime: FIXED_MTIME,
            uid: 0,
            gid: 0,
            uname: "",
            gname: "",
        }, (err) => {
            if (err) {
                reject(err);
                return;
            }
            resolve();
        });
        fileStream.on("error", reject);
        tarEntry.on("error", reject);
        fileStream.pipe(tarEntry);
    });
}
export function createTarStream(entries) {
    const pack = tar.pack();
    void (async () => {
        try {
            for (const entry of entries) {
                if (entry.type === "directory") {
                    await writeDirEntry(pack, entry);
                }
                else {
                    await writeFileEntry(pack, entry);
                }
            }
            pack.finalize();
        }
        catch (err) {
            pack.destroy(err);
        }
    })();
    return pack;
}
